core/ibft.go

```diff
@@ -0,0 +1,1134 @@
+	select {
+	case i.roundDone <- struct{}{}:
+	case <-ctx.Done():
+	}
+}
+
+// signalNewRCC notifies the sequence routine (RunSequence) that
+// a valid RCC for a higher round appeared
+func (i *IBFT) signalNewRCC(ctx context.Context, round uint64) {
+	select {
+	case i.roundCertificate <- round:
+	case <-ctx.Done():
+	}
+}
+
+type newProposalEvent struct {
+	proposalMessage *proto.Message
+	round           uint64
+}
+
+// signalNewProposal notifies the sequence routine (RunSequence) that
+// a valid proposal for a higher round appeared
+func (i *IBFT) signalNewProposal(ctx context.Context, event newProposalEvent) {
+	select {
+	case i.newProposal <- event:
+	case <-ctx.Done():
+	}
+}
+
+// watchForFutureProposal listens for new proposal messages
+// that are intended for higher rounds
+func (i *IBFT) watchForFutureProposal(ctx context.Context) {
+	var (
+		view      = i.state.getView()
+		height    = view.Height
+		nextRound = view.Round + 1
+
+		sub = i.messages.Subscribe(
+			messages.SubscriptionDetails{
+				MessageType: proto.MessageType_PREPREPARE,
+				View: &proto.View{
+					Height: height,
+					Round:  nextRound,
+				},
+				NumMessages: 1,
+				HasMinRound: true,
+			})
+	)
+
+	defer func() {
+		i.messages.Unsubscribe(sub.ID)
+
+		i.wg.Done()
+	}()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case round := <-sub.SubCh:
+			proposal := i.handlePrePrepare(&proto.View{Height: height, Round: round})
+			if proposal == nil {
+				continue
+			}
+
+			// Extract the proposal
+			i.signalNewProposal(
+				ctx,
+				newProposalEvent{proposal, round},
+			)
+
+			return
+		}
+	}
+}
+
+// watchForRoundChangeCertificates is a routine that waits
+// for future valid Round Change Certificates that could
+// trigger a round hop
+func (i *IBFT) watchForRoundChangeCertificates(ctx context.Context) {
+	defer i.wg.Done()
+
+	var (
+		view   = i.state.getView()
+		height = view.Height
+		round  = view.Round
+		quorum = i.backend.Quorum(height)
+
+		sub = i.messages.Subscribe(messages.SubscriptionDetails{
+			MessageType: proto.MessageType_ROUND_CHANGE,
+			View: &proto.View{
+				Height: height,
+				Round:  round + 1, // only for higher rounds
+			},
+			NumMessages: 1,
+			HasMinRound: true,
+		})
+	)
+
+	defer i.messages.Unsubscribe(sub.ID)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case round := <-sub.SubCh:
+			rcc := i.handleRoundChangeMessage(
+				&proto.View{
+					Height: height,
+					Round:  round,
+				},
+				quorum,
+			)
+			if rcc == nil {
+				continue
+			}
+
+			//	we received a valid RCC for a higher round
+			i.signalNewRCC(ctx, round)
+
+			return
+		}
+	}
+}
+
+// RunSequence runs the IBFT sequence for the specified height
+func (i *IBFT) RunSequence(ctx context.Context, h uint64) {
+	// Set the starting state data
+	i.state.clear(h)
+	i.messages.PruneByHeight(h)
+
+	i.log.Info("sequence started", "num", h)
+	defer i.log.Info("sequence done", "num", h)
+
+	for {
+		i.wg.Add(4)
```

Comment:
I feel to need to move this line immediately before starting go rountines.

Reply:
Sorry, I just wanted to say to you "How about moving this line after logging". It's clear that 4 routines are spawned after this. You can add 4 in a line 

```go
view := i.state.getView()
i.log.Info("round started", "round", view.Round)

i.wg.Add(4)
```

Reply:
Ah, I misunderstood 🤦‍♂️ 

Moved:

[a547203](https://github.com/0xPolygon/go-ibft/pull/21/commits/a547203a1ecf76e4cc3767fa9a114223152d8047)